### PR TITLE
Season of Docs: MicroPython Internals documentation Outline

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
As part of season of docs, I present this outline for discussion on content for MicroPython internals. I started a discussion thread on this two weeks ago [in the discussion forum]( https://forum.micropython.org/viewtopic.php?f=2&t=8328) .

I will appreciate any thoughts.

### CODE REVIEW

File: docs/micropythoninternals/index.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
+.. _micropython_internals:
+
+MicroPython Internals
+=====================
+
+This chapter covers a tour of MicroPython from the perspective of a developer.
```
Comment (SpotlightKid) [inline docs/micropythoninternals/index.rst:6]: Suggestion;

> from the perspective of a developer, working on the implementation of MicroPython itself.

File: docs/micropythoninternals/library.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
+.. _internals_library:
+
+Implementing a MicroPython Library
+==================================
+
+This section decribes using a simple walk-through example how to implement a
+new Micropython module that can be installed with ``upip``.
```
Comment (SpotlightKid) [inline docs/micropythoninternals/library.rst:7]: > a new MicroPython module implemented in C, [...]

**Comment by dlech** [comment]
This look really good. I'm particularly interested in the compiler bits since that seem to be the least-well documented area in my experience.

A few suggestions:
- This looks like it could be integrated into the existing `develop` folder instead of making a new `internals` section.
- "Implementing a MicroPython library" should probably discuss implementing in (Micro)Python vs. implementing in C.
- Related to porting, embedding MicroPython (e.g. as a library) also seems to be a common use case.

**Comment by nanjekyejoannah** [comment]
>     * "Implementing a MicroPython library" should probably discuss implementing in (Micro)Python vs. implementing in C.

I have a section on adding a core module(built in) is this what you mean?

**Comment by dlech** [comment]
> I have a section on adding a core module(built in) is this what you mean?

There can also be external C modules. https://docs.micropython.org/en/latest/develop/cmodules.html

**Comment by nanjekyejoannah** [comment]
> 
> 
> > I have a section on adding a core module(built in) is this what you mean?
> 
> There can also be external C modules. https://docs.micropython.org/en/latest/develop/cmodules.html

I also have a subsection "supporting C modules" under the core section. I may not have used the right phrase though.

**Comment by tve** [comment]
I believe something overlooked in this outline is the (or are there multiple) internal public API embodied by runtime.h and obj.h I think (see, after a dozen PRs I'm not even sure myself what Damien exactly refers to when he refers to "the public API").

Is a section on upip that useful? In the age of github I have yet to find a good workflow around it. Maybe it does show up prominently in tutorials?

Another topic is putting all the pieces together. For example, what all happens where in the code when you do something like "import foo; foo.bar()".

**Comment by nanjekyejoannah** [comment]
> Is a section on upip that useful? In the age of github I have yet to find a good workflow around it. Maybe it does show up prominently in tutorials?

I think my intention for this section is to show how to extend MicroPython with a third party module, since part of MicroPython development is also maintaining such modules. If we give incite to this, the process will be more transparent to e contributors.

**Comment by nanjekyejoannah** [comment]
> I believe something overlooked in this outline is the (or are there multiple) internal public API embodied by runtime.h and obj.h I think (see, after a dozen PRs I'm not even sure myself what Damien exactly refers to when he refers to "the public API").

Can we liken this to the CPython C API? Technically there is internal and public too.

**Comment by stinos** [comment]
> I believe something overlooked in this outline is the (or are there multiple) internal public API embodied by runtime.h and obj.h I think (see, after a dozen PRs I'm not even sure myself what Damien exactly refers to when he refers to "the public API").

runtime/obj is the meat of it, but essentially the public API consists of all non-static functions and macros in header files.

**Comment by tve** [comment]
Then there's also `dynruntime.h` for dynamically loaded modules...

**Comment by nanjekyejoannah** [comment]
Thanks all for the suggestions. I added some more suggestions.

**Review by nanjekyejoannah — COMMENTED** [review]
Thanks all for the suggestions. I edited the outline to address the suggestions.

**Comment by nanjekyejoannah** [comment]
Am waiting for consensus on whether this should move to under `develop`

**Comment by nanjekyejoannah** [comment]
Closing this since we came to an agreement sort of on the outline and the docs are being written

**Comment by nanjekyejoannah** [comment]
Reopening to revisit the outline discussion

**Pull request closed without merging**